src/python/pants/jvm/resolve/coursier_setup.py

```diff
@@ -185,7 +199,7 @@ def args(self, args: Iterable[str], *, wrapper: Iterable[str] = ()) -> tuple[str
             self.post_process_stderr,
             *wrapper,
             os.path.join(self.bin_dir, self.coursier.exe),
-            *args,
+            *map(shlex.quote, args),
```

Comment:
Hm, thanks for trying this. I think technically this may not be quite right. This appears to be building the appropriate arguments for a single command, and (as it was), would've been safe/fine to pass to a command like `subprocess.run(coursier.args(...))` (or Pants' equivalent). By quoting here, that command would see the wrong argument, e.g. instead of something like `subprocess.run(["coursier", "has space"])` it would be `subprocess.run(["coursier", "'has space'"])` and the command would receive those quotes literally, unintentionally. This potentially happens at the following (although maybe the arguments rarely need quoting, so tests don't fail):

https://github.com/pantsbuild/pants/blob/100c4c7ba8e14c6d99663749b4c26eb2bf602fad/src/python/pants/jvm/resolve/coursier_setup.py#L239-L243


The special thing about what happens in `prepare_jdk_environment` is the invocation is splatted into a shell script as one long text string, rather than invoking it with a Python-ful API directly (and those APIs can preserve the separation between arguments in the list).

Thus, I think the best way to fix this is close to what you have but just  in the other file, i.e.

```diff
- optionally_prefixed_coursier_args = [prefixed(arg) for arg in coursier.args(coursier_options)]
+ optionally_prefixed_coursier_args = [shlex.quote(prefixed(arg)) for arg in coursier.args(coursier_options)]
```

and restore this back to how it was.

Reply:
🤦🏽 thanks for catching Huon..

Reply:
Applying `shlex.quote` after `prefixed` seems to be causing the environment variable appended by `prefixed` to be interpreted literally.  I get an error like.
```
/bin/bash: ${PANTS_INTERNAL_ABSOLUTE_PREFIX}__coursier/coursier_post_process_stderr.py: No such file or directory
```


Would changing the order like this cause any other issues?
```python
optionally_prefixed_coursier_args = [prefixed(shlex.quote(arg)) for arg in coursier.args(coursier_options)]
```

Reply:
Ah, yes, the quoting will indeed transform it to something like `'${PANTS_...}` intentionally, so that the `$` comes through literally, but that's not what we want. 

I think your proposal is close, but will cause issues if a `__` prefixed arg needs quoting, e.g. `__something here` would become `'__something here'`. Previously this would transform to `${PANTS_INTERNAL_ABSOLUTE_PREFIX}__something here` but 
the `startswith` would be false with the quoted version and so it'll come out as just `'__something here'`, incorrectly.

So... maybe move it within `prefixed` and be careful about where one uses the quoted vs. unquoted values:

```python
def prefixed_and_quote(arg: str) -> str:
    quoted = shlex.quote(arg)
    if arg.startswith("__"):
        return f"${{PANTS_INTERNAL_ABSOLUTE_PREFIX}}{quoted}"
    else:
        return quoted
```

Reply:
I've updated with that approach, thanks for all the feedback!